### PR TITLE
core(driver): add waitForFCP timeout

### LINKED ISSUE #7174: Increased failures due to 60s timeout on PSI/LR
We've seen an increase in our error rate on the LR backend:
image
(The red line is our render error rate)

It appears to be related to our latency, which has also increased some.

Since we have our .timings data available, I viewed that, which points to the major problem:
image

The chart here is the 95th percentile of each of these timings. `lh:runner:auditing` always remains flat, regardless of percentile. `loadPage-defaultPass` is the only timing to make a big jump. (Also this is a little tricky because any run that errors did NOT report these timings... so hypothetically a gatherer that takes 45s would never be visible to us.)

The loadPage jump is certainly happening regardless.. Here's that one timing in isolation, with the full heatmap:
image

Looking at the [diff of LH changes that was in that push](https://github.com/GoogleChrome/lighthouse/compare/d3b95b2d6...965861083).. I suspect https://github.com/GoogleChrome/lighthouse/pull/6944 "core(driver): waitForFCP when tracing".. Plus we also know that NO_FCP is our most commonly seen LighthouseError, so I suspect more sites hitting the 35s maxWaitForLoad timeout means more hitting the 60s render timeout.

-----------------

**What can we do about this?**

@brendankenny mentioned a potential threshold on how long we'd hold out for FCP.
@patrickhulce wdyt?

### PR DESCRIPTION
**Pull request by patrickhulce**
**Summary**
Adds a `maxWaitForFcp` option that can optionally timeout the wait for FCP trigger earlier than overall load.

**Related Issues/PRs**
closes #7174

### CODE REVIEW

File: lighthouse-core/config/constants.js [hunk 1]
```diff
@@ -42,6 +42,7 @@ const throttling = {
 /** @type {LH.Config.Settings} */
 const defaultSettings = {
   output: 'json',
+  maxWaitForFcp: 30 * 1000,
```
Comment (patrickhulce) [inline lighthouse-core/config/constants.js]: open to other defaults, 20s maybe? not sure what's reasonable

Comment (paulirish) [reply lighthouse-core/config/constants.js]: 10s is fine IMO.

Comment (patrickhulce) [reply lighthouse-core/config/constants.js]: That'll probably be too low when using real throttling though, no? :/

how about 15?

**Comment by paulirish** [comment]
sure

File: lighthouse-core/gather/driver.js [hunk 2]
```diff
@@ -1027,19 +1039,22 @@ class Driver {
...
 
       /* eslint-disable max-len */
       if (typeof pauseAfterLoadMs !== 'number') pauseAfterLoadMs = DEFAULT_PAUSE_AFTER_LOAD;
       if (typeof networkQuietThresholdMs !== 'number') networkQuietThresholdMs = DEFAULT_NETWORK_QUIET_THRESHOLD;
       if (typeof cpuQuietThresholdMs !== 'number') cpuQuietThresholdMs = DEFAULT_CPU_QUIET_THRESHOLD;
       if (typeof maxWaitMs !== 'number') maxWaitMs = constants.defaultSettings.maxWaitForLoad;
+      if (typeof maxFCPMs !== 'number') maxFCPMs = constants.defaultSettings.maxWaitForFcp;
       /* eslint-enable max-len */
 
+      if (!waitForFCP) maxFCPMs = undefined;
```
Comment (brendankenny) [inline lighthouse-core/gather/driver.js]: it's a little confusing that there are two similar-but-different parameters for this, but this way does seem to make the most sense from both gather-runner and driver's perspectives, so I guess this is where we live now :)

File: lighthouse-core/gather/driver.js [hunk 3]
```diff
@@ -881,6 +888,11 @@ class Driver {
       return function() {
         log.verbose('Driver', 'loadEventFired and network considered idle');
       };
+    }).catch(err => {
+      // Throw the error in the cleanupFn so we still cleanup all our handlers.
```
Comment (brendankenny) [inline lighthouse-core/gather/driver.js]: this isn't getting easier to follow... :P

**Comment by brendankenny** [comment]
NO_FCP in the smoke tests, though, and that's for the 30s timeout? 😟

File: lighthouse-core/gather/driver.js (hunk 3 above)
Comment (patrickhulce) [reply lighthouse-core/gather/driver.js]: no, it's not, but it's very well tested now and ready for refactoring :)

**Comment by patrickhulce** [comment]
> NO_FCP in the smoke tests, though, and that's for the 30s timeout? 😟

yeah this makes me pretty nervous :/

**Comment by patrickhulce** [comment]
oh @brendankenny the FCP timeout was a legit issue with one of the smoketests having no content, now this fails early instead of letting the audits fail

**Comment by benschwarz** [comment]
I think this is likely to cause issues when "real" throttling is used… (but of course I have no idea). 
There'd be no drawback in setting this timeout to be the same as the max load timeout for devtools throttling, right? Should that be a default for that scenario?

**Comment by patrickhulce** [comment]
> I think this is likely to cause issues when "real" throttling is used…

Yeah this was my concern with 10s. I'm not sure I've seen beyond more than 15s, but maybe this is more of a breaking change from that perspective...

**Pull request merged**